bin/propolis-server/src/lib/serial/backend.rs

```diff
@@ -0,0 +1,397 @@
+    ) -> ClientHandle {
+        let mut inner = self.inner.lock().unwrap();
+        let id = inner.next_client_id();
+        let client =
+            Client { tx: read_tx, read_discipline: full_read_tx_discipline };
+
+        inner.clients.insert(id, client);
+        ClientHandle { id, backend: self.clone(), permissions }
+    }
+
+    /// Returns the contents of this backend's history buffer. See
+    /// [`HistoryBuffer::contents_vec`].
+    pub fn history_vec(
+        &self,
+        byte_offset: SerialHistoryOffset,
+        max_bytes: Option<usize>,
+    ) -> Result<(Vec<u8>, usize), super::history_buffer::Error> {
+        let inner = self.inner.lock().unwrap();
+        inner.buffer.contents_vec(byte_offset, max_bytes)
+    }
+
+    /// Returns the number of bytes that have ever been sent to this backend's
+    /// history buffer.
+    pub fn bytes_since_start(&self) -> usize {
+        self.inner.lock().unwrap().buffer.bytes_from_start()
+    }
+}
+
+impl Drop for ConsoleBackend {
+    fn drop(&mut self) {
+        let (tx, _rx) = oneshot::channel();
+        let done_tx = std::mem::replace(&mut self.done_tx, tx);
+        let _ = done_tx.send(());
+    }
+}
+
+mod migrate {
+    use propolis::migrate::{
+        MigrateCtx, MigrateSingle, MigrateStateError, PayloadOffer,
+    };
+
+    use crate::serial::history_buffer::migrate::HistoryBufferContentsV1;
+
+    use super::ConsoleBackend;
+
+    impl MigrateSingle for ConsoleBackend {
+        fn export(
+            &self,
+            _ctx: &MigrateCtx,
+        ) -> Result<propolis::migrate::PayloadOutput, MigrateStateError>
+        {
+            Ok(self.inner.lock().unwrap().buffer.export().into())
+        }
+
+        fn import(
+            &self,
+            mut offer: PayloadOffer,
+            _ctx: &MigrateCtx,
+        ) -> Result<(), MigrateStateError> {
+            let contents: HistoryBufferContentsV1 = offer.parse()?;
+            self.inner.lock().unwrap().buffer.import(contents);
+            Ok(())
+        }
+    }
+}
+
+/// Reads bytes from the supplied `source` and dispatches them to the clients in
+/// `inner`. Each backend is expected to spin up one task that runs this
+/// function.
+async fn read_task(
+    inner: Arc<Mutex<Inner>>,
+    source: Arc<dyn Source>,
+    mut done_rx: oneshot::Receiver<()>,
+) {
+    const READ_BUFFER_SIZE_BYTES: usize = 512;
+    let buf = SourceBuffer::new(propolis::chardev::pollers::Params {
+        poll_interval: std::time::Duration::from_millis(10),
+        poll_miss_thresh: 5,
+        buf_size: NonZeroUsize::new(READ_BUFFER_SIZE_BYTES).unwrap(),
+    });
+    buf.attach(source.as_ref());
+
+    let mut bytes = vec![0; READ_BUFFER_SIZE_BYTES];
+    loop {
+        let bytes_read = select! {
+            biased;
+
+            _ = &mut done_rx => {
+                return;
+            }
+
+            res = buf.read(bytes.as_mut_slice(), source.as_ref()) => {
+                res.unwrap()
```

Comment:
nit: do we expect that this will never be an error? it might be nice to `expect` this with a string explaining why...